lib/cn/hblock_builder.c

```diff
@@ -0,0 +1,327 @@
+#include "omf.h"
+#include "wbt_builder.h"
+#include "blk_list.h"
+
+struct hblock_builder {
+    struct mpool *mpool;
+    const struct cn *cn;
+    struct wbb *ptree;
+    unsigned int ptree_pgc;
+    uint32_t max_size;
+    enum hse_mclass_policy_age agegroup;
+};
+
+static unsigned int
+available_pgc(struct hblock_builder *bld)
+{
+    return (bld->max_size / PAGE_SIZE) - HBLOCK_HDR_PAGES - HLOG_PGC;
+}
+
+static void HSE_NONNULL(1)
+make_header(
+    struct hblock_hdr_omf *hdr,
+    const uint64_t min_seqno,
+    const uint64_t max_seqno,
+    const uint32_t num_kblocks,
+    const uint32_t num_vblocks,
+    const uint32_t num_vgroups,
+    const uint32_t ptree_pgc,
+    const uint32_t vblk_idx_adj_pgc,
+    const struct key_obj *const min_pfx,
+    const struct key_obj *const max_pfx)
+{
+    /* If we have one, we must have both */
+    assert(((min_pfx && max_pfx) || (!min_pfx && !max_pfx)));
+
+    omf_set_hbh_magic(hdr, HBLOCK_HDR_MAGIC);
+    omf_set_hbh_version(hdr, HBLOCK_HDR_VERSION);
+    omf_set_hbh_min_seqno(hdr, min_seqno);
+    omf_set_hbh_max_seqno(hdr, max_seqno);
+    omf_set_hbh_num_kblocks(hdr, num_kblocks);
+    omf_set_hbh_num_vblocks(hdr, num_vblocks);
+    omf_set_hbh_num_vgroups(hdr, num_vgroups);
+    omf_set_hbh_hlog_off_pg(hdr, HBLOCK_HDR_PAGES);
+    omf_set_hbh_hlog_len_pg(hdr, HLOG_PGC);
+    omf_set_hbh_ptree_data_off_pg(hdr, HBLOCK_HDR_PAGES + HLOG_PGC);
+    omf_set_hbh_ptree_data_len_pg(hdr, ptree_pgc);
+    omf_set_hbh_vblk_idx_adj_off_pg(hdr, HBLOCK_HDR_PAGES + HLOG_PGC + ptree_pgc);
+    omf_set_hbh_vblk_idx_adj_len_pg(hdr, vblk_idx_adj_pgc);
+
+    if (max_pfx) {
+        unsigned int max_pfx_len = 0;
+
+        /* Make sure there is enough room in the space allocated */
+        static_assert(BYTE_ALIGN(HBLOCK_HDR_LEN - 2 * HSE_KVS_PFX_LEN_MAX) >= sizeof(*hdr),
+            "Not enough space in the hblock to store the min and max prefixes");
+
+        max_pfx_len = key_obj_len(max_pfx);
+        assert(max_pfx_len <= HSE_KVS_PFX_LEN_MAX);
+
+        omf_set_hbh_max_pfx_off(hdr, BYTE_ALIGN(HBLOCK_HDR_LEN - 2 * HSE_KVS_PFX_LEN_MAX));
+        omf_set_hbh_max_pfx_len(hdr, max_pfx_len);
+
+        key_obj_copy((void *)hdr + omf_hbh_max_pfx_off(hdr), HSE_KVS_PFX_LEN_MAX, NULL, max_pfx);
+    }
+
+    if (min_pfx) {
+        unsigned int min_pfx_len = 0;
+
+        min_pfx_len = key_obj_len(min_pfx);
+        assert(min_pfx_len <= HSE_KVS_PFX_LEN_MAX);
+
+        omf_set_hbh_min_pfx_off(hdr, BYTE_ALIGN(HBLOCK_HDR_LEN - HSE_KVS_PFX_LEN_MAX));
+        omf_set_hbh_min_pfx_len(hdr, min_pfx_len);
+
+        key_obj_copy((void *)hdr + omf_hbh_min_pfx_off(hdr), HSE_KVS_PFX_LEN_MAX, NULL, min_pfx);
+    }
+}
+
+merr_t
+hbb_add_ptomb(
+    struct hblock_builder *const bld,
+    const struct key_obj *const kobj,
+    const void *const kmd,
+    uint kmd_len,
+    struct key_stats *const stats)
+{
+    merr_t err;
+    bool   added = false;
+
+    assert(stats->nptombs > 0);
+
+    err = wbb_add_entry(
```

Comment:
I hate this alignment from clang.. Wasting lines for no good reason.

Reply:
I just copied that line of code from somewhere else. Flattened in next revision.